Music Keyboard basic structure added

sparsh0204 opened the pull request:
Basic structure for showing clicked notes

walterbender commented:
Good start. The order of the notes in the table should probably be flipped.

sparsh0204 commented:
Some major changes and improvements involve-
- Display table improved.
- Play using keyboard.
- Press duration captured.
- Press captured even if multiple keys are pressed at once.
Play and export function might not be working.

walterbender commented:
(1) Maybe the button for returning to the keyboard should look like a keyboard instead of a circle?
(2) Play seems to be broken for both the keyboard and the table.

js/widgets/musickeyboard.js:201 Uncaught TypeError: Cannot read property 'style' of null
    at HTMLTableCellElement.cell.onclick (js/widgets/musickeyboard.js:201)

walterbender commented:
(3) Still would like the table to be sorted the other way.
(4) The timing stuff needs some work... I ended up with a 0/2 note.

sparsh0204 commented:
1) I will change the icon. Circle was applied just for testing. 
2) Play and export and not working as I had to change how everything works inside musical keyboard. I will work on them once other things are approved.
3) I will change it back.
4) I have to apply check for edge cases.
Is everything else fine or something need any improvement or change

sparsh0204 commented:
Several changes- 
1) Sorting order changed to what it was before.
2) Added icon for keyboard.
3) Play button is now working.
4) Export button is noe working.
5) Clear button is now working.
6) Added graphics for when user click on play.
Not added- 
* Stop play button.

walterbender commented:
We have been miscommunicating about the order of the notes. Regardless of the order in the widget clamp, the notes on the keyboard should be sorted from lowest (left) to highest (right). In the table, it should be lowest (bottom) to highest (top).

walterbender commented:
Did the variable duration go away? I can only get 1/8 notes now.

walterbender commented:
Do we want to add a key (below, like a space bar?) for inserting rests?

sparsh0204 commented:
By lowest to highest do you mean frequency.
Yes I fixed the bug where a user can get a 0/8 note, user now can get any note rounded off to the nearest 1/8 as we discussed.
No currently there isn't any key for rests. Should I insert a key in keyboard layout mapped with space bar for the same(In table it can be represented same as phrase maker.).

walterbender commented:
In theory, we could try to replicate the press and release of a piano key and hence avoid the need for rests, but that presumes the user knows how to play piano and has their composition well in mind as they play. Maybe we need two modes of recording: one with timing on and one with timing off. In at least the latter mode, having a space bar for rests would be nice.

sparsh0204 commented:
This is something I have been thinking about. A mode which records event the rest part. Something like this - 
User clicks on advanced mode(assumed name of feature) then the widget will record everything like duration of press, time from the start when the button is pressed and can even record if a note is pressed in the duration of press of another note along with rest and then the user clicks on stop advanced mode and can play the recorded note.

walterbender commented:
An advanced record sounds good. But lets wrap up all the other things first.

sparsh0204 commented:
I am still a bit confused about the sorting order.

walterbender commented:
Sorting should always be by frequency. On the keyboard, lowest frequency to the left, highest to the right. On the grid, lowest frequency at the bottom, highest frequency at the top.

walterbender commented:
![Screenshot from 2019-08-03 18-29-22](https://user-images.githubusercontent.com/3996641/62417426-a3812480-b61c-11e9-87a1-1aac14e32a1a.png)
This is still flipped from what it should be. Do4 should be at the bottom.

walterbender commented:
Thanks for fixing the table.

FWIW:

(1) The playback from the table is a bit fast. Definitely not 1/8 notes.
(2) Playback from the keyboard is broken:
js/widgets/musickeyboard.js:384 Uncaught TypeError: Cannot read property 'style' of null
    at MusicKeyboard.playAll (js/widgets/musickeyboard.js:384)
    at HTMLTableCellElement.cell.onclick (js/widgets/musickeyboard.js:243)

sparsh0204 commented:
I improved the playback time.
But I was unable to reproduce the error about playback in the keyboard. It is working fine on my laptop no matter how I try. I also checked the code logic it also looks ok.

walterbender commented:
I'll test today.

walterbender commented:
Works much better. And I cannot reproduce the playback bug.

walterbender commented:
I merged these changes. We can add the other changes in another PR. (rests, timing, editing note values....)
